pkg/apis/pipeline/v1alpha1/gcs_resource.go

```diff
@@ -0,0 +1,172 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"flag"
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+)
+
+var (
+	gsutilImage              = flag.String("gsutil-image", "override-with-gsutil-image:latest", "The container image containing gsutil")
+	gcsSecretVolumeMountPath = "/var/secret"
+)
+
+// GCSResource is a GCS endpoint from which to get artifacts which is required
+// by a Build/Task for context (e.g. a archive from which to build an image).
+type GCSResource struct {
+	Name           string               `json:"name"`
+	Type           PipelineResourceType `json:"type"`
+	Location       string               `json:"location"`
+	TypeDir        bool                 `json:"typeDir"`
+	DestinationDir string               `json:"destinationDir"`
+	//Secret holds a struct to indicate a field name and corresponding secret name to populate it
+	Secrets []SecretParam `json:"secrets"`
+}
+
+// NewGCSResource creates a new GCS resource to pass to knative build
+func NewGCSResource(r *PipelineResource) (*GCSResource, error) {
+	if r.Spec.Type != PipelineResourceTypeStorage {
+		return nil, fmt.Errorf("GCSResource: Cannot create a GCS resource from a %s Pipeline Resource", r.Spec.Type)
+	}
+	var location string
+	var locationSpecified, dir bool
+
+	for _, param := range r.Spec.Params {
+		switch {
+		case strings.EqualFold(param.Name, "Location"):
+			location = param.Value
+			if param.Value != "" {
+				locationSpecified = true
+			}
+		case strings.EqualFold(param.Name, "Dir"):
+			dir = true // if dir flag is present then its a dir
+		}
+	}
+
+	if !locationSpecified {
+		return nil, fmt.Errorf("GCSResource: Need Location to be specified in order to create GCS resource %s", r.Name)
+	}
+	return &GCSResource{
+		Name:     r.Name,
+		Type:     r.Spec.Type,
+		Location: location,
+		TypeDir:  dir,
+		Secrets:  r.Spec.SecretParams,
+	}, nil
+}
+
+// GetName returns the name of the resource
+func (s GCSResource) GetName() string {
+	return s.Name
+}
+
+// GetType returns the type of the resource, in this case "storage"
+func (s GCSResource) GetType() PipelineResourceType {
+	return PipelineResourceTypeStorage
+}
+
+// GetParams get params
+func (s *GCSResource) GetParams() []Param { return []Param{} }
+
+// GetSecretParams returns the resource secret params
+func (s *GCSResource) GetSecretParams() []SecretParam { return s.Secrets }
+
+// Replacements is used for template replacement on an GCSResource inside of a Taskrun.
+func (s *GCSResource) Replacements() map[string]string {
+	return map[string]string{
+		"name":     s.Name,
+		"type":     string(s.Type),
+		"location": s.Location,
+	}
```

Comment:
what about dir?

Reply:
`Dir` is decided by controller and not replacement template. So did not add that key

Reply:
hm kk - it looks in the docs like ppl can specify dir explicitly:

> `dir`: represents whether the blob storage is a directory or not. By default storage artifact is considered not a directory.

so i feel like it would make sense to provide here? but i cant imagine why anyone would want to template with it so maybe it's fine!